### PR TITLE
Validate against json schemas

### LINKED ISSUE #5: Implement validation against json schemas

### PR DESCRIPTION
**Pull request by jiyoontbd**
closes #5

### CODE REVIEW

File: protocol/src/main/kotlin/Playground.kt
Comment (jiyoontbd) [inline protocol/src/main/kotlin/Playground.kt:1]: to remove before merging

File: protocol/src/main/resources/person.schema.json
Comment (jiyoontbd) [inline protocol/src/main/resources/person.schema.json:1]: to remove before merging

File: README.md [hunk 1]
```diff
@@ -6,6 +6,9 @@ This repo contains 3 jvm packages:
 * [`/http-server`](./http-server) - A configurable implementation of the [tbdex http api draft specification](https://github.com/TBD54566975/tbdex-protocol/blob/main/rest-api/README.md)
 
 ## tbDEX Schemas
+> **Warning**  
+> Until `tbdex` is stable, do not sync schemas into `tbdex-kt`
+
```
Comment (ethan-tbd) [inline README.md:11]: changes had to be made to the schemas in order to work with tbdex-kt, syncing would revert those changes

File: protocol/src/main/kotlin/models/Message.kt [hunk 2]
```diff
@@ -118,6 +97,34 @@ sealed class Message {
...
+
+      // validate message structure
+      Validator.validate(messageJson, "message")
+
+      val dataJson = messageJson.getJSONObject("data")
+      val kind = messageJson.getJSONObject("metadata").getString("kind")
+
+      // validate specific message data (Rfq, Quote, etc)
+      Validator.validate(dataJson, kind)
+    }
```
Comment (ethan-tbd) [inline protocol/src/main/kotlin/models/Message.kt:127]: potentially make these private functions?

File: protocol/src/main/kotlin/models/Resource.kt [hunk 3]
```diff
@@ -110,8 +86,36 @@ sealed class Resource {
...
+      
+      // validate message structure
+      Validator.validate(resourceJson, "resource")
+
+      val dataJson = resourceJson.getJSONObject("data")
+      val kind = resourceJson.getJSONObject("metadata").getString("kind")
+      
+      // validate specific resource data
+      Validator.validate(dataJson, kind)
+    }
```
Comment (ethan-tbd) [inline protocol/src/main/kotlin/models/Resource.kt:119]: same as [this comment](https://github.com/TBD54566975/tbdex-kt/pull/39/files#r1355912570)

File: protocol/src/main/resources/message.schema.json [hunk 4]
```diff
@@ -5,16 +5,16 @@
       "type": "object",
       "properties": {
         "from": {
-          "$ref": "https://tbdex.io/definitions.json#/definitions/did",
```
Comment (ethan-tbd) [inline protocol/src/main/resources/message.schema.json:8]: including the `https://tbdex.io/` led to `FileNotFoundException`

File: protocol/src/main/kotlin/Json.kt [hunk 5]
```diff
@@ -1,14 +1,61 @@
...
 
 const val dateTimeFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX"
 
+/**
+ * Custom Jackson JSON serializer for converting a [TypeID] object to a JSON string.
+ * This serializer is used to customize the JSON serialization of [TypeID] instances.
+ *
+ * @see TypeID
+ */
+class TypeIDToStringSerializer : JsonSerializer<TypeID>() {
```
Comment (phoebe-lew) [inline protocol/src/main/kotlin/Json.kt:25]: ❤️ love u for this

File: protocol/src/main/kotlin/Validator.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+ * Utility for validating JSON payloads against predefined schemas.
+ */
+object Validator {
+  private val schemaMap = mutableMapOf<String, Schema>()
+  private const val RESOLUTION_SCOPE_URL = "classpath:/"
+
+  /**
+   * Initializes the validator by loading schemas for messages and resources.
+   */
+  init {
```
Comment (phoebe-lew) [inline protocol/src/main/kotlin/Validator.kt:19]: cleannn, nice!

**Review by phoebe-lew — APPROVED** [review]

**Review by tomdaffurn — APPROVED** [review]

File: protocol/src/main/kotlin/Validator.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import models.MessageKind
```
[the remark this answers is not in the document]
Comment (jiyoontbd) [reply protocol/src/main/kotlin/Validator.kt:1]: i'd like to do namespacing in a separate PR! created issue [here](https://github.com/TBD54566975/tbdex-kt/issues/50)

File: protocol/src/main/kotlin/models/Message.kt (hunk 2 above)
Comment (jiyoontbd) [reply protocol/src/main/kotlin/models/Message.kt:127]: we use these in Message.validate() in MessageTest.kt... maybe not good to keep a method public purely for the sake of testing tho

File: protocol/src/main/kotlin/models/Resource.kt (hunk 3 above)
Comment (jiyoontbd) [reply protocol/src/main/kotlin/models/Resource.kt:119]: same here - use Resource.validate() in ResourceTst.kt (pushing this up now)

File: protocol/src/main/resources/message.schema.json (hunk 4 above)
Comment (jiyoontbd) [reply protocol/src/main/resources/message.schema.json:8]: see tbdex change here: https://github.com/TBD54566975/tbdex/pull/166

File: protocol/src/main/kotlin/Validator.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+   * @throws Exception if validation fails, including a list of validation errors.
+   */
+
+  fun validate(jsonMessage: JSONObject, schemaName: String) {
+    try {
+      val schema = schemaMap[schemaName] ?: throw Exception("No schema with name $schemaName exists")
+      schema.validate(jsonMessage)
+    } catch (e: ValidationException) {
+      val errorList = collectValidationErrors(e)
+      throw Exception("Validation failed, errors: $errorList")
```
[the remark this answers is not in the document]
Comment (jiyoontbd) [reply protocol/src/main/kotlin/Validator.kt]: done

File: protocol/src/main/kotlin/models/Message.kt [hunk 9]
```diff
@@ -118,6 +97,34 @@ sealed class Message {
...
+     * @throws Exception if the payload does not conform to the expected structure or data schema.
+     */
+    fun validate(payload: String) {
+      val messageJson = JSONObject(payload)
+
+      // validate message structure
+      Validator.validate(messageJson, "message")
+
+      val dataJson = messageJson.getJSONObject("data")
+      val kind = messageJson.getJSONObject("metadata").getString("kind")
```
[the remark this answers is not in the document]
Comment (jiyoontbd) [reply protocol/src/main/kotlin/models/Message.kt:122]: no worries! this is an excellent point. looks like we're using JsonNode in parse() and JSONObject in validate(). would like to align on using 1 Json ... thing and extract out the common bits cc @ethan-tbd

**Pull request merged**